### PR TITLE
DEX-513: merge requests

### PR DESCRIPTION
**Pull request by naknomum**
## Pull Request Overview

- Implements **Merge Request** - when a user wants to merge 2+ individuals, but does not have edit permission on _all individuals_, this gives other stakeholders a window of time to cancel that request (otherwise merge will go through after deadline passes)
- Full details outlined in _Merge API Design Doc_
- Request deadline handled by task in **Celery** (deadline currently _two weeks_)
- Notifications are sent to stakeholders when request is initiated

---

**Review Notes**
- A _Merge request_ is referenced by a "request id" which is actually the ID of the corresponding `AsyncResult` celery object
- This ID is used to query information about the request, as well as for stakeholders to vote on it
- A single vote to _block_ the request cancels it, which **revokes** the celery task, preventing the merge
- If the state of any involved Individual or their Encounters has changed significantly (from its initial state when the merge was requested) when the deadline arrives, the merge will not happen. (This is done by recording a _checksum_ of the key attributes of the Individuals involved at the time of the request.  If the checksum has changed when it is time to merge, it will fail.)

**REST API Updates**

A _merge request_ is initiated via API the exact same way a _merge_ is, but happens when the requesting user **does not have edit privileges on all** encounters of _all_ individuals.  Note: the user must have privilege on _at least one_ of the encounters.  If they have _zero_, the request will generate a 403.

```
POST /api/v1/individuals/TARGET_INDIVIDUAL_GUID/merge

  [ FROM_INDIV_1, .... FROM_INDIV_n ] 
```
The **response**, however, varies and includes both the _request id_ and deadline, as well as a list of _blocking encounters_ (those which the user does not have sufficient privileges on):
```
{
    "merge_request": true,
    "blocking_encounters": [ "e71a8a06-bb35-451b-8360-e571ed3b9951" ],
    "request_id": "9bc6212a-3ab0-4274-8441-bc86f5e22bfb",
    "deadline": "2021-11-11T23:30:25"
}
```
Information **about a merge request** can be obtained via this GET API call:
```
GET /api/v1/individuals/merge_request/9bc6212a-3ab0-4274-8441-bc86f5e22bfb

{
    "_secondsToDeadline": 1206617.571826,
    "_serverTime": "2021-11-12T07:31:39.736777+00:00",
    "_deadlinePolicyDays": 14,
    "eta": "2021-11-26T06:41:57.308587+00:00",
    "priority": 6,
    "request": {
        "id": "9bc6212a-3ab0-4274-8441-bc86f5e22bfb",
        "name": "app.modules.individuals.tasks.execute_merge_request",
        . . .
    },
    . . .
}
```

And also to **vote** on a merge request via POST.  (Accepted values are `allow` and `block`.)
```
POST /api/v1/individuals/merge_request/9bc6212a-3ab0-4274-8441-bc86f5e22bfb
{
    "vote":  "allow"
}
```

### CODE REVIEW

File: app/modules/encounters/models.py [hunk 1]
```diff
@@ -125,6 +125,16 @@ def add_annotation(self, annotation):
             with db.session.begin(subtransactions=True):
                 self.annotations.append(annotation)
 
+    def _merge_request_hash(self):
```
Comment (hwindsor) [inline app/modules/encounters/models.py:128]: This is not used in the encounter models code, only in the individuals models code, hence it should not be 'private', ie no preceding _

File: app/modules/individuals/models.py [hunk 2]
```diff
@@ -269,6 +321,118 @@ def _consolidate_social_groups(self, source_individual):
...
+
+    # does the actual work of setting up celery task to execute this merge
+    # NOTE: this does not do any notification of users; see merge_request_from()
+    def _merge_request_init(self, individuals, parameters=None):
+        from app.modules.individuals.tasks import execute_merge_request
+        from datetime import datetime, timedelta
+
+        if not individuals or not isinstance(individuals, list) or len(individuals) < 1:
+            msg = f'merge request passed invalid individuals: {individuals}'
+            AuditLog.backend_fault(log, msg, self)
```
Comment (hwindsor) [inline app/modules/individuals/models.py:337]: Why is this a backend fault? THis is the request we're sending to the backend, It's our fault, or probably the users, in other words houston or frontend fault

File: app/modules/individuals/resources.py [hunk 3]
```diff
@@ -515,3 +531,106 @@ def get(self, individual):
             except HoustonException as ex:
                 abort(ex.status_code, ex.message)
             return send_file(image_path, attachment_filename='individual_image.jpg')
+
+
+# TESTING ONLY
```
Comment (hwindsor) [inline app/modules/individuals/resources.py:536]: Shouldn't be leaving unused test code in the main program

File: app/modules/individuals/resources.py [hunk 4]
```diff
@@ -515,3 +531,106 @@ def get(self, individual):
...
+
+
+@api.route('/merge_request/<uuid:task_id>', defaults={'vote': None}, doc=False)
+@api.route('/merge_request/<uuid:task_id>/<string:vote>')
+class IndividualMergeRequestByTaskId(Resource):
+    """
+    Details of merge request, or register a vote on it
+    """
+
+    def get(self, task_id, vote):
```
Comment (hwindsor) [inline app/modules/individuals/resources.py:565]: Why is the GET API being used to POST a vote?
Surely the GET should return the status of the Merge request and POST should be used to vote on it

File: app/modules/individuals/tasks.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    log_id = f'<execute_merge_request {self.request.id}>'
+    log.info(
+        f'{log_id} initiated for Individual {target_individual_guid} (from {from_individual_ids}; {parameters})'
+    )
+    all_individuals = Individual.validate_merge_request(
+        target_individual_guid, from_individual_ids, parameters
+    )
+    if not all_individuals:
+        msg = f'{log_id} failed validation'
+        AuditLog.backend_fault(log, msg)
```
Comment (hwindsor) [inline app/modules/individuals/tasks.py:28]: again, why a backend fault?

File: app/modules/individuals/tasks.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+log = logging.getLogger(__name__)
+
+
+@celery.task(
+    bind=True,
+    # autoretry_for=(requests.exceptions.RequestException,),
+    # default_retry_delay=3,
+    # max_retries=10,
+)
+def execute_merge_request(self, target_individual_guid, from_individual_ids, parameters):
```
Comment (hwindsor) [inline app/modules/individuals/tasks.py:15]: The general pattern in other tasks is to do the minimum in the the task function as possible and have the functionality implemented in a model method. 
I'd suggest you should follow this pattern here too.

File: app/modules/notifications/models.py [hunk 7]
```diff
@@ -67,6 +72,24 @@ class NotificationChannel(str, enum.Enum):
         'email_digest_content_template': 'collaboration_edit_request_digest.jinja2',
         'mandatory_fields': {'collaboration_guid'},
     },
+    NotificationType.merge_request: {
```
Comment (hwindsor) [inline app/modules/notifications/models.py:75]: I'm thinking this should be an individual_merge_request to:
1 be unambiguous 
2 Not require rework if we ever decide we want to merge something else

File: app/modules/notifications/models.py [hunk 8]
```diff
@@ -85,6 +108,16 @@ def __init__(self, sender, multiple=False):
     def set_collaboration(self, collab):
         self.data['collaboration_guid'] = collab.guid
 
+    def set_merge_request(self, individuals, encounters, request_data):
```
Comment (hwindsor) [inline app/modules/notifications/models.py:111]: likewise, set_individual_merge_request

File: tests/modules/individuals/resources/test_merge.py [hunk 9]
```diff
@@ -107,39 +99,155 @@ def test_merge_permissions(
...
+    # invalid id (404)
+    response = individual_res_utils.get_or_vote_individuals(
+        flask_app_client,
+        contributor_1,  # just needs to not be anon
+        bad_id,
+        expected_status_code=404,
+    )
+
+    # now we need real data
+    sighting1, encounter1 = individual_utils.simple_sighting_encounter(
```
Comment (hwindsor) [inline tests/modules/individuals/resources/test_merge.py:177]: OK Now you can start using the asset group created one :)

**Comment by hwindsor** [comment]
Generally looks good. I like how the functionality has been broken down into different blocks in the models code, makes it very easy to read. 
A few comments need addressing, the most important one is 
Why are you using the GET API to POST a vote? 
There may be some really good fundamental reason for this but right now I'm really not seeinng it.

File: app/modules/encounters/models.py (hunk 1 above)
Comment (naknomum) [reply app/modules/encounters/models.py:128]: done.

File: app/modules/individuals/models.py (hunk 2 above)
Comment (naknomum) [reply app/modules/individuals/models.py:411]: oh yeah; i was abusing `backend_fault` ... did a bunch of fixing up of that now and am only using them where its certain the edm was involved.

File: app/modules/notifications/models.py (hunk 7 above)
Comment (naknomum) [reply app/modules/notifications/models.py:75]: agreed; i have done this as well as method change you mention below.

File: app/modules/individuals/resources.py (hunk 4 above)
Comment (naknomum) [reply app/modules/individuals/resources.py]: made **vote** now a POST

File: app/modules/individuals/tasks.py (hunk 6 above)
Comment (naknomum) [reply app/modules/individuals/tasks.py:14]: gotcha.  [done](https://github.com/WildMeOrg/houston/pull/365/commits/7da861d91c8db39c6e936a955a7baad927ac27e3).

File: app/modules/individuals/models.py [hunk 10]
```diff
@@ -231,15 +287,129 @@ def merge_from(self, *source_individuals, sex=None, primary_name=None):
...
+
+    # scrubs notifications etc, once a merge has completed
+    @classmethod
+    def merge_request_cleanup(cls, req_id):
+        log.debug(
+            f'[{req_id}] merge_request_cleanup (notifications, etc) not yet implemented'
+        )
+
+    @classmethod
+    def merge_request_celery_task(
```
Comment (hwindsor) [inline app/modules/individuals/models.py:371]: Celery task code can be tested (by mocking them to circumvent the need for celery) 
See 
https://github.com/WildMeOrg/houston/blob/main/tests/modules/asset_groups/resources/test_create_asset_group.py#L423-L428

File: app/modules/individuals/models.py [hunk 11]
```diff
@@ -231,15 +287,129 @@ def merge_from(self, *source_individuals, sex=None, primary_name=None):
...
+
+        # notify users that merge has happened
+        #   NOTE request_data here may need some altering depending on what final templates look like
+        #   also unclear who *sender* will be, so that may need to be passed
+        request_data = {
+            'id': cel_task.request.id,
+            'from_individual_ids': from_individual_ids,
+            'merge_outcome': 'deadline',
+        }
+        Individual.merge_request_notify(
```
Comment (hwindsor) [inline app/modules/individuals/models.py:410]: surely this is a notification of the merge itslef, not the request? I'd say that the word request here is misleading

File: app/modules/notifications/models.py [hunk 12]
```diff
@@ -85,6 +108,16 @@ def __init__(self, sender, multiple=False):
     def set_collaboration(self, collab):
         self.data['collaboration_guid'] = collab.guid
 
+    def set_individual_merge_request(self, individuals, encounters, request_data):
```
Comment (hwindsor) [inline app/modules/notifications/models.py:111]: likewise if this is used for the request and the complete notification, the word request here is misleading.

File: app/modules/individuals/models.py [hunk 13]
```diff
@@ -231,15 +287,129 @@ def merge_from(self, *source_individuals, sex=None, primary_name=None):
...
+    # - just a light wrapper to _merge_request_init()
+    def merge_request_from(self, source_individuals, parameters=None):
+        res = self._merge_request_init(source_individuals, parameters)
+        Individual.merge_request_notify([self] + source_individuals, res)
+        return res
+
+    # note: for merge_complete individuals will only contain target individual, but
+    #   request_data should contain key from_individual_ids
+    @classmethod
+    def merge_request_notify(cls, individuals, request_data, notif_type=None):
```
Comment (hwindsor) [inline app/modules/individuals/models.py:309]: likewise, ditch request

File: app/modules/individuals/models.py [hunk 14]
```diff
@@ -231,15 +287,129 @@ def merge_from(self, *source_individuals, sex=None, primary_name=None):
...
+                current_user,
+                owner,
+                owners[owner]['individuals'],
+                owners[owner]['encounters'],
+                request_data,
+                notif_type,
+            )
+
+    @classmethod
+    def _merge_request_notify_user(
```
Comment (hwindsor) [inline app/modules/individuals/models.py:342]: and again

File: app/modules/individuals/models.py [hunk 15]
```diff
@@ -231,15 +287,129 @@ def merge_from(self, *source_individuals, sex=None, primary_name=None):
...
+        log_id = f'<execute_merge_request {cel_task.request.id}>'
+        log.info(
+            f'{log_id} initiated for Individual {target_individual_guid} (from {from_individual_ids}; {parameters})'
+        )
+        all_individuals = Individual.validate_merge_request(
+            target_individual_guid, from_individual_ids, parameters
+        )
+        if not all_individuals:
+            msg = f'{log_id} failed validation'
+            AuditLog.backend_fault(log, msg)
```
Comment (hwindsor) [inline app/modules/individuals/models.py:385]: backend fault?

Comment (naknomum) [reply app/modules/individuals/models.py:385]: fixed, + one other

**Review by hwindsor — APPROVED** [review]
LGTM, presuming request droppage which you've agreed to

**Review by hwindsor — APPROVED** [review]
Great, time for this one to go in and you to take the big step in the next PR for the test changes :)

**Pull request merged**